Allow directly writing snapshots on top of existing files

CompuIves opened the pull request:
# Reason for This PR

We're exploring diff snapshots, but found out that our underlying filesystem (btrfs) doesn't perform well with the `rebase-snap` tool due to performance issues with finding holes in files (related: https://github.com/firecracker-microvm/firecracker/issues/2948). Initially we tried using another syscall (`fiemap`), but that also was too slow (10s for 4GB). Because of this, we cannot apply diff snapshots quickly on top of a base snapshot.

Instead, we're now taking another approach where we directly save the diff snapshot on top of the base memory. Firecracker already supports this, because it only writes changed blocks. This has worked out well for us, and this might be interesting to merge into upstream.

There's no issue opened for this right now, and it might well be that this change cannot be merged into Firecracker, but if you're interested I'd be happy to make the required changes to make it fit for the repo!

## Description of Changes

Instead of truncating the memory file we write to, we write on top of it.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [ ] The issue which led to this PR has a clear conclusion.
- [ ] This PR follows the solution outlined in the related issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes follow the [Runbook for Firecracker API changes][2].
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

acatangiu commented:
My 2c is this is a valid use-case that `firecracker` could support _in addition_ to existing behavior of creating _new_ snapshot containing just _the diff bits_.

There are two different use-cases here:
1. (existing) the user is interested in the _diff pages_ of memory, the _diff snapshot_ - for example, multiple layers of "diffs" can used in a distributed system to build snapshots at varying stages in the VM's lifetime, while minimizing the amount of "data" to distribute across the system.
2. (what this PR proposes) the user is interested in a single snapshot instance, but wants to accelerate mutations to it - take snapshots as the guest workload makes progress (but only interested in latest snapshot), so instead of creating slow(er) full snapshots, directly create fast(er) diff snapshots on top of each other.

I think there's value in both use-cases, and the proposed changes in this PR seem to support both of them :+1:

CompuIves commented:
It seems like the tests are failing on test coverage, should I change a value here?

luminitavoicu commented:
@CompuIves, thanks for your contribution! Your changes increase ARM coverage. Please rebase your PR and update the coverage value for the failing tests.

luminitavoicu commented on src/vmm/src/persist.rs:
```diff
@@ -327,14 +327,20 @@ fn snapshot_memory_to_file(
     let mem_size_mib = mem_size_mib(vmm.guest_memory());
-    file.set_len((mem_size_mib * 1024 * 1024) as u64)
-        .map_err(|e| MemoryBackingFile("set_length", e))?;
+    let expected_size = (mem_size_mib * 1024 * 1024) as u64;
```
Not necessarily in scope of this PR, but I would like us to have a constant for `1024 * 1024`.

luminitavoicu reviewed (commented):
Looks good, I like the idea! Could you add an integration test for this, as well?

CompuIves commented:
Rebased the PR, but I'm not sure what the code coverage is now with the changes that came from the rebase. Should I wait for a test run?

CompuIves commented:
Added integration test as well!

luminitavoicu reviewed (commented):
Looks good! Please make sure that all of your commits are signed. 
The coverage test is failing, you will have to update the numbers once again. Also take a look at the errors output from the python style test.

CompuIves commented:
Addressed the feedback, squashed the commits, signed it and updated coverage! I think the tests should now pass, though I did see other perf tests fail, so that's something we should check.

dianpopa commented:
Hi!

I do like the proposal! However, I am concerned about the possibility of a breaking change since I do expect that users of the `rebase-snap` tool to have the expectation to obtain diff layers containing dirty memory since *last* retrieval.
Shouldn't we let rebase-snap tool users know that from now in order to obtain a diff layer they need to make sure to remove any previous (name conflicting) snapshots?

roypat commented:
> Hi!
> 
> I do like the proposal! However, I am concerned about the possibility of a breaking change since I do expect that users of the `rebase-snap` tool to have the expectation to obtain diff layers containing dirty memory since _last_ retrieval. Shouldn't we let rebase-snap tool users know that from now in order to obtain a diff layer they need to make sure to remove any previous (name conflicting) snapshots?

Mh, yes, for successive diff snapshots, this essentially does "rebase-snap" functionality at snapshot-taking-time, we might want to add a note about that somewhere. For full snapshots however, this "fixes" the scenario where currently if a full snapshot from which the microvm is loaded is being used as the target of a full/diff snapshot, it just gets zeroed out.
